### PR TITLE
[SE-0362] Add `enableUpcomingFeature` and `enableExperimentalFeature` Swift settings

### PR DESCRIPTION
**Pull request by DougGregor**
Implement the SwiftPM part of [SE-0362 "Piecemeal adoption of upcoming language improvements"](https://github.com/apple/swift-evolution/blob/main/proposals/0362-piecemeal-future-features.md).

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

File: Sources/PackageDescription/BuildSettings.swift [hunk 1]
```diff
@@ -296,6 +296,53 @@ public struct SwiftSetting: Encodable {
...
+    /// be used in a target without affecting its dependencies. An unknown
+    /// future feature will be ignored by the implementation.
+    ///
+    /// - Since: First available in PackageDescription 5.7.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the future feature, e.g., ConciseMagicFile.
+    ///   - condition: A condition that restricts the application of the build
+    /// setting.
+    public static func enableFutureFeature(
```
Comment (tomerd) [inline Sources/PackageDescription/BuildSettings.swift:316]: does this need `@available(_PackageDescription, introduced: 5.7)`?

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageDescription/BuildSettings.swift (hunk 1 above)
Comment (DougGregor) [reply Sources/PackageDescription/BuildSettings.swift:316]: I bet it does, thank you!

**Comment by DougGregor** [comment]
I believe from the code that these flags are being treat as "unsafe", which we didn't want for `enableFutureFeature`. But it seems rather invasive to change.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: Sources/PackageLoading/PackageBuilder.swift [hunk 2]
```diff
@@ -927,6 +927,30 @@ public final class PackageBuilder {
...
+                case .c, .cxx, .linker:
+                    throw InternalError(
+                        "only Swift supports experimental features")
+
+                case .swift:
+                    decl = .OTHER_SWIFT_FLAGS
+                }
+
+                values = _values.precedeElements(
+                    with: "-enable-experimental-feature")
```
Comment (benrimmington) [inline Sources/PackageLoading/PackageBuilder.swift:953]: I don't think `precedeElements(with:)` is needed, if the new APIs are implemented like the existing `.define` API, which only stores a single value.

```swift
case .enableExperimentalFeature(let value):
    values = ["-enable-experimental-feature", value]
    switch setting.tool {
```

**Comment by hassila** [comment]
Just a friendly ping: Out of curiosity, is this going to be part of 5.8? We plan to prepare all our repos for Swift 6 features now and try to decide whether to wait for a 5.x with this nice feature enabled or whether to use the various other mechanisms to enable them.

**Comment by DougGregor** [comment]
Thanks for the ping! Yes, I plan to get this into 5.8.

**Comment by DougGregor** [comment]
(I still need to rename "future" to "upcoming", per the proposal's review discussion)

**Comment by hassila** [comment]
Super, thanks! Then we'll look forward to testing it out and plan accordingly!

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@neonichu how's my implementation  look?

**Review by neonichu — APPROVED** [review]

**Pull request merged**